src/aeso_parser.erl

```diff
@@ -211,12 +211,20 @@ letdef() -> choice(valdef(), fundef()).
 valdef() ->
     ?RULE(pattern(), tok('='), body(), {letval, [], _1, _3}).
 
-fundef() ->
+fundef() -> choice(unguarded_fundef(), guarded_fundef()).
+
+unguarded_fundef() ->
     choice(
     [ ?RULE(id(), args(),                   tok('='), body(), {letfun, get_ann(_1), _1, _2, type_wildcard(get_ann(_1)), _4})
     , ?RULE(id(), args(), tok(':'), type(), tok('='), body(), {letfun, get_ann(_1), _1, _2, _4, _6})
     ]).
 
+guarded_fundef() ->
+    choice(
+    [ ?RULE(id(), args(),                   tok('|'), expr(), tok('='), body(), {letfun, get_ann(_1), _1, _2, type_wildcard(get_ann(_1)), _4, _6})
+    , ?RULE(id(), args(), tok(':'), type(), tok('|'), expr(), tok('='), body(), {letfun, get_ann(_1), _1, _2, _4, _6, _8})
```

Comment:
Might be also good to make `guard/0` a separate parser

Reply:
Why do you think that guards should have comma separated exprs? Isn't it better if a single expr is used and boolean operators are used when needed?

Reply:
Because later we might want to add support for introducing variables in guards. Just like in Haskell you can do
```
f x | y <- x + x, y > 10 = 10
```